[AERIE-1826] Accumulate activity/configuration instantiation errors

pcrosemurgy opened the pull request:
* **Tickets addressed:** AERIE-1826
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description

Following #140 this PR properly accumulates instantiation errors for:
- Unconstructable (unserializable) arguments
- Extraneous arguments
- Missing arguments

Previously these failure cases were handled by a single exception and would fail-fast, resulting in a maximum of one argument failure per activity/configuration. With this PR an `InvalidArgumentsException` reports all unconstructable/extraneous/missing arguments that were encountered.

## Verification

<details>
<summary>Expand</summary>

```gql
  getActivityEffectiveArguments
  (
    missionModelId: 1,
    activityTypeName: "BakeBananaBread",
    activityArguments: { tbSugar: "hi", extra: 1 }
  ) {
    success
    errors
    arguments
  }
```
→
```json
{
  "data": {
    "getActivityEffectiveArguments": {
      "success": false,
      "errors": {
        "unconstructableArguments": [
          {
            "name": "tbSugar",
            "failure": "Expected integral number, got StringValue[value=hi]"
          }
        ],
        "missingArguments": [
          "tbSugar",
          "glutenFree"
        ],
        "extraneousArguments": [
          "extra"
        ]
      },
      "arguments": {
        "temperature": 350
      }
    }
  }
}
```

See future work section for serialization/structuring of these exceptions within HTTP/GQL responses.

</details>

## Documentation

Updated `FooActivityMapper` example within `aerie.wiki`'s `Activity-Mappers.md` (https://github.com/NASA-AMMOS/aerie/wiki/Activity-Mappers#generated-activity-mapper-example).

## Future work

Captured in https://jira.jpl.nasa.gov/browse/AERIE-1968:

Work is needed to decide where these exceptions should be serialized/structured. For some endpoints it may be enough to just call `toString()` while others may benefit from a structured response. Ultimately the UI's needs may drive these decisions.

Specifically, the `validatePlan` endpoint returns a `reason` message with that is currently unstructured. This field and the `simulation_dataset.reason` column within our Merlin DB should likely be structured. This will also require looking at all instances of `writer.failWith()`:
```
merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/SynchronousSimulationAgent.java
merlin-worker/src/main/java/gov/nasa/jpl/aerie/merlin/worker/MerlinWorkerAppDriver.java
scheduler-server/src/main/java/gov/nasa/jpl/aerie/scheduler/server/services/SynchronousSchedulerAgent.java
```

pcrosemurgy committed:
Refactor common `MapperMethodMaker` control flow

mattdailis commented on merlin-framework-processor/src/main/java/gov/nasa/jpl/aerie/merlin/processor/generator/MapperMethodMaker.java:
```diff
@@ -175,15 +177,107 @@ public final MethodSpec makeGetValidationFailuresMethod() {
+        .endControlFlow()
+        .addCode("\n");
+
+    if (shouldExpectArguments) {
+      mb = mb
+          .addStatement(
+              "$L.throwIfAny()",
+              "unconstructableArgsExBuilder");
+    }
+
+    mb = mb
+        .addStatement(
+            "$L.throwIfAny()",
+            "extraneousArgsExBuilder")
+        .addCode("\n");
```
🔶 I'm wondering what we gain or lose from separating these three illegal argument exception types (Unconstructable, Extraneous, and Missing). 

One thing we lose is that we only ever throw one of these three, but the same activity could have all three simultaneously.

I'm wondering if collecting all of this information into one exception, and throwing it at the end, would allow us to both provide the maximum amount of feedback to users, and avoid cluttering the MissionModel interface with several related exception types.

pcrosemurgy replied:
Very good point. I'm cool with going this direction! I'll rework things today to just throw one exception.

pcrosemurgy replied:
This change has now been made :)

pcrosemurgy committed:
Remove unused facade exceptions

Removes:
- `UnconfigurableMissionModelException`
- `MissionModelContractException`

pcrosemurgy committed:
Replace `MissingArgumentsException` with `InvalidArgumentsException`

Missing arguments exceptions are now replaced by more general invalid arguments exceptions.
`InvalidArgumentsException.Builder` facilitates the accumulation of extraneous, unconstructable, and missing arguments.
This allows `instantiate()` methods to not fail-fast when one of these invalid arguments is encountered.
This change propagates from generated mapper code all the way up to HTTP bindings.

pcrosemurgy committed:
Update `aerielander.jar`

pcrosemurgy committed:
Add `InvalidArgumentsException`

dyst5422 reviewed (approved):
I like the approach here! 👍

mattdailis commented on merlin-server/src/main/java/gov/nasa/jpl/aerie/merlin/server/services/MissionModelService.java:
```diff
@@ -32,57 +31,47 @@ Map<String, ActivityType> getActivityTypes(String missionModelId)
   // TODO: Provide a finer-scoped validation return type. Mere strings make all validations equally severe.
   List<String> validateActivityArguments(String missionModelId, SerializedActivity activity)
-  throws NoSuchMissionModelException, TaskSpecType.UnconstructableTaskSpecException, MissingArgumentsException;
+      throws NoSuchMissionModelException, InvalidArgumentsException;
```
Rogue indentation

pcrosemurgy replied:
Fixed, thanks!

pcrosemurgy replied:
Marking as resolved to merge PR :)

mattdailis commented on merlin-server/src/test/java/gov/nasa/jpl/aerie/merlin/server/models/MissionModelTest.java:
```diff
@@ -102,8 +102,7 @@ public void shouldNotGetActivityTypeForNonexistentActivityType() {
     public void shouldInstantiateActivityInstance()
-    throws MissionModelFacade.NoSuchActivityTypeException, MissionModelFacade.MissionModelContractException,
-           TaskSpecType.UnconstructableTaskSpecException, MissingArgumentsException
+        throws MissionModelFacade.NoSuchActivityTypeException, InvalidArgumentsException
```
Rogue indentation

pcrosemurgy replied:
Also fixed :)

mattdailis commented on merlin-sdk/src/main/java/gov/nasa/jpl/aerie/merlin/protocol/types/UnconstructableArgumentException.java:
```diff
@@ -0,0 +1,11 @@
+package gov.nasa.jpl.aerie.merlin.protocol.types;
+
+public final class UnconstructableArgumentException extends Exception {
```
🔵 I suspect we could do away with this exception (since we're throwing it then immediately catching it), but I won't insist on that for this PR.

pcrosemurgy replied:
Oh, yeah that try/catch is kind of annoying. It’s needed since the only way to get a failure string is to use `getSuccessOrThrow()`, which only accepts a function that throws an exception.

For an example:
```java
    for (final var entry : arguments.entrySet()) {
      try {
        switch (entry.getKey()) {
          case "x":
            template.x = this.mapper_x.deserializeValue(entry.getValue())
                .getSuccessOrThrow(failure -> new UnconstructableArgumentException("x", failure));
            break;
          case "y":
            template.y = this.mapper_y.deserializeValue(entry.getValue())
                .getSuccessOrThrow(failure -> new UnconstructableArgumentException("y", failure));
            break;
          case "vecs":
            template.vecs = this.mapper_vecs.deserializeValue(entry.getValue())
                .getSuccessOrThrow(failure -> new UnconstructableArgumentException("vecs", failure));
            break;
          default:
            invalidArgsExBuilder.withExtraneousArgument(entry.getKey());
        }
      } catch (final UnconstructableArgumentException e) {
        invalidArgsExBuilder.withUnconstructableArgument(e.parameterName, e.failure);
      }
    }
```

Using the `getSuccessOrThrow()` defined here: https://github.com/NASA-AMMOS/aerie/blob/develop/merlin-framework/src/main/java/gov/nasa/jpl/aerie/merlin/framework/Result.java?plain=1#L112

The `throws Throws` requires us to throw an exception if we want to cache/save `failure`. We could maybe change/add to `Result` to have a method that allows us to not `throws Throws`. Maybe I'm missing some other simpler option, open to other ideas :)

mattdailis reviewed (approved):
Looks good to me - a couple formatting comments, and one musing about UnconstructableArgumentException that can be ignored if desired.
